umbraco/Umbraco-CMS
# Data type usage reporting and deletion warning

Displays a report when viewing a data type to show where it is used: Document types, media types, member types and optionally for which property editors.  This feature also adds a new deletion prompt when deleting a data type to let you know exactly where it is used before you agree to delete it.

![image](https://user-images.githubusercontent.com/1742685/66544671-84ef5a80-eb84-11e9-924e-a8b678d8f638.png)

![image](https://user-images.githubusercontent.com/1742685/66544763-b9fbad00-eb84-11e9-85eb-59b259e06b42.png)

## Changes

* data type ui updated to support content app layout - this means moving some stuff around since each view is now responsible for the parts of the data type layout
* data type angular updated to use more modern apis (controller as, etc...)
* report data is only requested when the releations tab is active
* new IDataTypeService.GetReferences
* new DataTypeController.GetReferences endpoint for angular
* Fixes navigating in angular when navigating via an html local url (there was a bug in this)
* updated delete dialog/controller to show where the data type is used, this now uses more modern apis  (controller as, etc...)

## Testing

* Ensure unit tests all pass
* Review code
* Ensure data type creation & editing all works as expected
* Ensure the relations report is working and is accurate when clicking the relations tab
* Ensure you can click through to the related content type items
* Ensure the relations delete report is working and is accurate when trying to delete a data type and make sure you can still delete the data type
